pymeasure/instruments/redpitaya/redpitaya_scpi.py

```diff
@@ -0,0 +1,329 @@
+log = logging.getLogger(__name__)
+log.addHandler(logging.NullHandler())
+
+
+class DigitalChannelP(Channel):
+    """ A digital line of the P type"""
+
+    direction = Channel.control(
+        "DIG:PIN:DIR? DIO{ch}_P", "DIG:PIN:DIR %s,DIO{ch}_P",
+        """ Control a digital line to the given direction (str either 'IN' or 'OUT')""",
+        validator=truncated_discrete_set,
+        values=['IN', 'OUT'],
+    )
+
+    state = Channel.control(
+        "DIG:PIN? DIO{ch}_P", "DIG:PIN DIO{ch}_P,%d",
+        """ Control the state of the line (bool)""",
+        validator=strict_discrete_set,
+        map_values=True,
+        values={True: 1, False: 0},
+    )
+
+
+class DigitalChannelN(Channel):
+    """ A digital line of the N type"""
+
+    direction = Channel.control(
+        "DIG:PIN:DIR? DIO{ch}_N", "DIG:PIN:DIR %s,DIO{ch}_N",
+        """ Control a digital line to the given direction (str either 'IN' or 'OUT')""",
+        validator=truncated_discrete_set,
+        values=['IN', 'OUT'],
+    )
+
+    state = Channel.control(
+        "DIG:PIN? DIO{ch}_N", "DIG:PIN DIO{ch}_N,%d",
+        """ Control the state of the line (bool)""",
+        validator=strict_discrete_set,
+        map_values=True,
+        values={True: 1, False: 0},
+    )
+
+
+class DigitalChannelLed(Channel):
+    """ A LED digital line (Output only)"""
+
+    state = Channel.control(
+        "DIG:PIN? LED{ch}", "DIG:PIN LED{ch},%d",
+        """ Control the state of the led (bool)""",
+        validator=strict_discrete_set,
+        map_values=True,
+        values={True: 1, False: 0},
+    )
+
+
+class AnalogInputSlowChannel(Channel):
+    """ A slow analog input channel"""
+
+    voltage = Channel.measurement(
+        "ANALOG:PIN? AIN{ch}",
+        """ Measure the voltage on the corresponding analog input channel, range is [0, 3.3]V""",
+    )
+
+
+class AnalogOutputSlowChannel(Channel):
+    """ A slow analog output channel"""
+
+    voltage = Channel.setting(
+        "ANALOG:PIN AOUT{ch}, %f",
+        """ Set the voltage on the corresponding analog input channel, range is [0, 1.8]V""",
+        validator=truncated_range,
+        values=[0, 1.8],
+    )
+
+
+class AnalogInputFastChannel(Channel):
+
+    gain = Instrument.control(
+        "ACQ:SOUR{ch}:GAIN?", "ACQ:SOUR{ch}:GAIN %s",
+        """Control the gain of the selected fast analog input either 'LV' or 'HV' (see jumpers on boards)
+
+        'LV' set the returned values in the range [-1, 1]V and 'HV' in the range [-20, 20]V
+        """,
+        validator=strict_discrete_set,
+        values=['LV', 'HV'],
+    )
+
+    def get_data_from_ascii(self) -> np.ndarray:
+        self.write(f"ACQ:SOUR{self.id}:DATA?")
+        data_str = self.read()
+        return np.fromstring(data_str.strip('{}').encode(), sep=',')
+
+    def get_data_from_binary(self) -> np.ndarray:
+        """ Get the data in """
+        self.write(f"ACQ:SOUR{self.id}:DATA?")
+        self.parent.adapter.read_bytes(1)
```

Comment:
You can just call `read_bytes`, which calls the parent's `read_bytes`.
That way the parent does not have to be the instrument, which has an adapter, but could be another channel.
More importantly, your channel does not have to know how the parent is designed (with an adapter or not).

Reply:
done